pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -45,12 +47,108 @@ func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
+
+// takes a list of pods sorted based on their resource requirements compared by CPU and memory.
+
+func (p *PodPacker) packSortedPods(pods []*v1.Pod, instanceTypes []*instanceType) ([]*Packings, error) {
+	// Start with the smallest instance type and the biggest pod check how
+	// many pods can we fit, go to the next bigger type and check if we can fit
+	// more pods. Compare pods packed on all and select the instance type with
+	// highest pod count.
+	podsPacked := map[*v1.Pod]bool{}
+	packings := make([]*Packings, 0)
+	for start, pod := range pods {
+		if podsPacked[pod] {
+			continue
+		}
+
+		prevPacking := make([]*v1.Pod, 0)
+		instanceTypesSelected := make([]*instanceType, 0)
+		// Go to the next instance type see how many more pods can we fit?
+		for _, instance := range instanceTypes {
+			// from start index loop through all the pods and check how many
+			// pods we can fit on each instance type
+			packings, err := calculateBestPackingOption(instance, start, pods, podsPacked)
+			if err != nil {
+				zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instance.name, err)
+				continue
+			}
+			zap.S().Info("packings are ", len(packings))
+			// If the pods packed are the same as before, this instance type can be
+			// considered a backup option in case we get ICE
+			if len(packings) == len(prevPacking) &&
+				podsMatch(prevPacking, packings) {
+				instanceTypesSelected = append(instanceTypesSelected, instance)
+			} else if len(packings) > len(prevPacking) {
+				// If pods packed are more than compared to what we did last time,
+				// consider using this instance type
+				prevPacking = packings
+				instanceTypesSelected = []*instanceType{instance}
+			}
+		}
+		// checked all instance type and found no packing option
+		if len(prevPacking) == 0 {
+			return nil, fmt.Errorf("no instance type found for packing %d pods", len(pods))
+		}
+		// keep a track of pods we have packed
+		for _, pod := range prevPacking {
+			podsPacked[pod] = true
+		}
+		instanceOptions := []string{}
+		for _, instanceType := range filterInstancesBasedOnCost(instanceTypesSelected) {
+			instanceOptions = append(instanceOptions, instanceType.name)
+		}
+		zap.S().Info("Instance Options are %v", instanceOptions)
+		packings = append(packings, &Packings{prevPacking, instanceOptions})
+	}
+	return packings, nil
+}
+
+func podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podSeen := map[*v1.Pod]struct{}{}
+	for _, pod := range first {
+		podSeen[pod] = struct{}{}
+	}
+	for _, pod := range second {
+		if _, exist := podSeen[pod]; !exist {
+			return false
+		}
+	}
+	return true
+}
+
+func calculateBestPackingOption(instanceType *instanceType, start int, podList []*v1.Pod, packed map[*v1.Pod]bool) ([]*v1.Pod, error) {
+	packing := make([]*v1.Pod, 0)
+	for i := start; i < len(podList); i++ {
+		pod := podList[i]
+		if packed[pod] {
+			continue
+		}
+		cpu := calculateCPURequested(pod)
```

Comment:
I find it a touch awkward to rely on private functions in other files. This is equivalent to using the "protected" keyword in java, which I think should generally be used sparingly in all languages. I typically treat lower case functions as file scoped, even if you can technically see them in another file. This helps the code be well encapsulated and makes refactors more straight forward.

Reply:
I agree for the most part, I had instance logic and pod resources all in same file and it grew to be too long, so had to separate the functionality into 3 different files. In this case, since it's being used by both the files `packing.go` and `podresources.go`, had to prefer DRY.